core/src/main/java/jenkins/security/AgentSecretAction.java

```diff
@@ -0,0 +1,62 @@
+package jenkins.security;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import hudson.model.Computer;
+import hudson.model.UnprotectedRootAction;
+import hudson.slaves.SlaveComputer;
+import jenkins.model.Jenkins;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+public class AgentSecretAction implements UnprotectedRootAction {
+    private static final Logger LOGGER = Logger.getLogger(AgentSecretAction.class.getName());
+
+    @Override
+    public String getUrlName() {
+        return "agent-secret";
+    }
+
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+    public void doGet(StaplerRequest req, StaplerResponse rsp, @QueryParameter String nodeName) throws IOException {
+        Jenkins jenkins = Jenkins.get();
+        jenkins.checkPermission(Computer.AGENT_SECRET);
```

Comment:
This permission check is wrong. You have to check the permission on the agent itself not in jenkins. A user might have the permission on one agent but not on another.